### ISSUE TITLE
CnnSentenceDataSetIterator with CollectionLabeledSentenceProvider for multi-label setting

### ISSUE THREAD

**Issue by mtothak** [deeplearning4j/deeplearning4j#5001]
#### Issue Description

Use of CnnSentenceDataSetIterator together with CollectionLabeledSentenceProvider in a multi-label setting

**Comment by AlexDBlack** [comment]
This was discussed in gitter. Basically: the goal is to have CnnSentenceDataSetIterator be usable for multi-label tasks.

**Comment by tschut** [comment]
Implemented this in my project by creating `MultiLabeledSentenceProvider`, based on `LabeledSentenceProvider`:
```
public interface MultiLabeledSentenceProvider {

    /**
     * Are there more sentences/documents available?
     */
    boolean hasNext();

    /**
     * @return Pair: sentence/document text and label(s)
     */
    Pair<String, List<String>> nextSentence();

    /**
     * Reset the iterator - including shuffling the order, if necessary/appropriate
     */
    void reset();

    /**
     * Return the total number of sentences, or -1 if not available
     */
    int totalNumSentences();

    /**
     * Return the list of labels - this also defines the class/integer label assignment order
     */
    List<String> allLabels();

    /**
     * Equivalent to allLabels().size()
     */
    int numLabelClasses();
}
```

And then adjusting CnnSentenceDataSetIterator to basically loop over the label and set appropriate 1's in the labels array it produces. I didn't do this for the various LabelAwareIterators because I'm not using them, but it should be similar.

**Comment by AlexDBlack** [comment]
> And then adjusting CnnSentenceDataSetIterator to basically loop over the label and set appropriate 1's in the labels array it produces.

One problem here is that we really need a MultiDataSetIterator to support multiple labels, not a DataSetIterator.
So while your MultiLabeledSentenceProvider seems fine, you can't just put those into a single labels array, unless you are doing binary multi-label learning.

Other than that, the interface seems fine.

**Comment by tschut** [comment]
Admittedly I'm a bit confused about the exact specifics of _binary multi-label_. Say I have a list of _n_ topics and each sentence can be classified by having [0-n] topics, this would be binary multi-label classification, right? There's 1 class (topic), n labels, and each label is either there or it isn't. So basically it's the same as single-class multi-label as I understand it.

Of course, if you have a multi-class multi-label problem, my suggestion above will not be enough.

**Comment by AlexDBlack** [comment]
> Say I have a list of n topics and each sentence can be classified by having [0-n] topics, this would be binary multi-label classification, right? 

Yes, that's binary multi-label. You can put them all in one label INDArray and use sigmoid + xent. Hence you can use DataSetIterator for that.
